File: typescript/leetcodehard/sudokuSolver.ts
/*
https://leetcode.com/problems/sudoku-solver/

status - completed
*/

import { logp } from "../utils/logger";

function solveSudoku(board: string[][]): void {
  function isValid(row: number, col: number, ch: string): boolean {
    for (let i = 0; i < 9; i++) {
      if (board[row][i] === ch) return false;
      if (board[i][col] === ch) return false;
      const boxRow = 3 * Math.floor(row / 3) + Math.floor(i / 3);
      const boxCol = 3 * Math.floor(col / 3) + (i % 3);
      if (board[boxRow][boxCol] === ch) return false;
    }
    return true;
  }

  function backtrack(): boolean {
    for (let row = 0; row < 9; row++) {
      for (let col = 0; col < 9; col++) {
        if (board[row][col] !== ".") continue;

        for (let d = 1; d <= 9; d++) {
          const ch = String(d);
          if (!isValid(row, col, ch)) continue;

          board[row][col] = ch;
          if (backtrack()) return true;
          board[row][col] = ".";
        }
        return false;
      }
    }
    return true;
  }

  backtrack();
}

function boardToString(board: string[][]): string {
  return board.map((row) => row.join("")).join("|");
}

const board = [
  ["5", "3", ".", ".", "7", ".", ".", ".", "."],
  ["6", ".", ".", "1", "9", "5", ".", ".", "."],
  [".", "9", "8", ".", ".", ".", ".", "6", "."],
  ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
  ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
  ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
  [".", "6", ".", ".", ".", ".", "2", "8", "."],
  [".", ".", ".", "4", "1", "9", ".", ".", "5"],
  [".", ".", ".", ".", "8", ".", ".", "7", "9"],
];

logp(`before solving: ${boardToString(board)}`);
solveSudoku(board);
logp(`after solving: ${boardToString(board)}`);

const expected =
  "534678912" +
  "672195348" +
  "198342567" +
  "859761423" +
  "426853791" +
  "713924856" +
  "961537284" +
  "287419635" +
  "345286179";
const actual = board.map((row) => row.join("")).join("");
logp(`expected solved rows joined ${expected} actual ${actual} match ${expected === actual}`);

File: python/leetcodehard/sudoku_solver.py
"""
https://leetcode.com/problems/sudoku-solver/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def solveSudoku(self, board):
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]
        empty_cells = []

        for r in range(9):
            for c in range(9):
                value = board[r][c]
                if value == ".":
                    empty_cells.append((r, c))
                else:
                    rows[r].add(value)
                    cols[c].add(value)
                    boxes[self.box_index(r, c)].add(value)

        self.backtrack(board, empty_cells, 0, rows, cols, boxes)

    def box_index(self, r, c):
        return (r // 3) * 3 + (c // 3)

    def backtrack(self, board, empty_cells, pos, rows, cols, boxes):
        if pos == len(empty_cells):
            return True

        r, c = empty_cells[pos]
        box = self.box_index(r, c)

        for digit in "123456789":
            if digit in rows[r] or digit in cols[c] or digit in boxes[box]:
                continue

            board[r][c] = digit
            rows[r].add(digit)
            cols[c].add(digit)
            boxes[box].add(digit)

            if self.backtrack(board, empty_cells, pos + 1, rows, cols, boxes):
                return True

            board[r][c] = "."
            rows[r].remove(digit)
            cols[c].remove(digit)
            boxes[box].remove(digit)

        return False


if __name__ == "__main__":
    test_input_1 = [
        ["5", "3", ".", ".", "7", ".", ".", ".", "."],
        ["6", ".", ".", "1", "9", "5", ".", ".", "."],
        [".", "9", "8", ".", ".", ".", ".", "6", "."],
        ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
        ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
        ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
        [".", "6", ".", ".", ".", ".", "2", "8", "."],
        [".", ".", ".", "4", "1", "9", ".", ".", "5"],
        [".", ".", ".", ".", "8", ".", ".", "7", "9"],
    ]
    Solution().solveSudoku(test_input_1)
    Solution().log.info("Output is %s", test_input_1)

    test_input_2 = [
        ["."] * 9,
        [".", ".", ".", ".", ".", ".", ".", "3", "."],
        [".", ".", "2", ".", ".", ".", ".", ".", "."],
        [".", ".", ".", ".", ".", ".", "5", ".", "4"],
        [".", ".", ".", ".", ".", "1", ".", ".", "."],
        [".", "4", ".", ".", ".", ".", ".", ".", "."],
        [".", ".", ".", "1", ".", ".", ".", ".", "."],
        [".", ".", ".", ".", ".", ".", ".", ".", "."],
        [".", ".", ".", ".", ".", ".", ".", ".", "."],
    ]
    Solution().solveSudoku(test_input_2)
    Solution().log.info("Output is %s", test_input_2)
